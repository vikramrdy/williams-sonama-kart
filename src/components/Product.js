import { GridListTileBar, IconButton, makeStyles } from "@material-ui/core";
import BrokenImageIcon from "@material-ui/icons/BrokenImage";
import InfoIcon from "@material-ui/icons/Info";
import PropTypes from "prop-types";
import React from "react";

const useStyles = makeStyles(() => ({
  icon: {
    color: "rgba(255, 255, 255, 0.54)",
  },
  noImage: {
    alignContent: "center",
    color: "dimGray",
    height: 300,
    width: "100%",
  },
  image: {
    width: "100%",
  },
}));

const Product = ({ data }) => {
  const classes = useStyles();

  const { name, priceRange, lowPrice, highPrice, hero } = data;

  return (
    <>
      {hero?.href ? (
        <img src={hero?.href} alt={name} className={classes.image} />
      ) : (
        <BrokenImageIcon
          data-testid="brokenImage"
          className={classes.noImage}
        />
      )}
      <GridListTileBar
        title={name}
        subtitle={
          priceRange ? (
            <span>
              {`$${lowPrice}`} - {`$${highPrice}`}
            </span>
          ) : (
            "Price - TBD"
          )
        }
        actionIcon={
          <IconButton
            aria-label={`info about ${name}`}
            className={classes.icon}
          >
            <InfoIcon />
          </IconButton>
        }
      />
    </>
  );
};

Product.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string,
    priceRange: PropTypes.any,
    lowPrice: PropTypes.number,
    highPrice: PropTypes.number,
    hero: PropTypes.shape({
      href: PropTypes.string,
    }),
  }).isRequired,
};

export default Product;
