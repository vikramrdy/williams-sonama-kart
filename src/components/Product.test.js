import { render } from "@testing-library/react";
import React from "react";

import Product from "./Product";

describe("Product Component", () => {
  it("should match the snapshot", () => {
    const props = {
      data: {
        name: "Heather Taylor Home Gingham & Stripe Bath Towels",
        priceRange: { selling: { high: 34.5, low: 24.5 } },
        lowPrice: 24.5,
        highPrice: 34.5,
        hero: {
          href:
            "https://assets.weimgs.com/weimgs/rk/images/wcm/products/202116/0258/heather-taylor-home-gingham-stripe-bath-towel-2-m.jpg",
        },
      },
    };
    const { asFragment } = render(<Product {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should contain the product name", () => {
    const props = {
      data: {
        name: "Heather Taylor Home Gingham & Stripe Bath Towels",
        priceRange: { selling: { high: 34.5, low: 24.5 } },
        lowPrice: 24.5,
        highPrice: 34.5,
        hero: {
          href:
            "https://assets.weimgs.com/weimgs/rk/images/wcm/products/202116/0258/heather-taylor-home-gingham-stripe-bath-towel-2-m.jpg",
        },
      },
    };
    const { getByText } = render(<Product {...props} />);
    expect(
      getByText(/Heather Taylor Home Gingham & Stripe Bath Towels/i)
    ).toBeInTheDocument();
  });

  it("should not render price if price range is not present", () => {
    const props = {
      data: {
        name: "Heather Taylor Home Gingham & Stripe Bath Towels",
        hero: {
          href:
            "https://assets.weimgs.com/weimgs/rk/images/wcm/products/202116/0258/heather-taylor-home-gingham-stripe-bath-towel-2-m.jpg",
        },
      },
    };
    const { getByText } = render(<Product {...props} />);
    expect(getByText(/Price - TBD/i)).toBeInTheDocument();
  });

  it("should render broken image if image is not present", () => {
    const props = {
      data: {
        name: "Heather Taylor Home Gingham & Stripe Bath Towels",
      },
    };
    const { getByTestId } = render(<Product {...props} />);
    expect(getByTestId("brokenImage")).toBeTruthy();
  });
});
