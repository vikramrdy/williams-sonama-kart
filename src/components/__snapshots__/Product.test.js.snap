// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Component should match the snapshot 1`] = `
<DocumentFragment>
  <img
    alt="Heather Taylor Home Gingham & Stripe Bath Towels"
    class="makeStyles-image-3"
    src="https://assets.weimgs.com/weimgs/rk/images/wcm/products/202116/0258/heather-taylor-home-gingham-stripe-bath-towel-2-m.jpg"
  />
  <div
    class="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
  >
    <div
      class="MuiGridListTileBar-titleWrap MuiGridListTileBar-titleWrapActionPosRight"
    >
      <div
        class="MuiGridListTileBar-title"
      >
        Heather Taylor Home Gingham & Stripe Bath Towels
      </div>
      <div
        class="MuiGridListTileBar-subtitle"
      >
        <span>
          $24.5 - $34.5
        </span>
      </div>
    </div>
    <div
      class="MuiGridListTileBar-actionIcon"
    >
      <button
        aria-label="info about Heather Taylor Home Gingham & Stripe Bath Towels"
        class="MuiButtonBase-root MuiIconButton-root makeStyles-icon-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
